### PR TITLE
Client: README Overhaul

### PR DESCRIPTION
**Pull request by holgerd77**
Ok, this PR completely overhauls the README of the client by doing some major restructuring, deleting various outdated sections as well as adding some new ones. I have the impression this can still be the first iteration of several more to come, but it should prepare some sufficient ground to have everything up-to-date and complete enough for a client release next week. 😄

Best to (re-)view by looking at the rendered [README](https://github.com/ethereumjs/ethereumjs-monorepo/blob/522397708a655e0ff015429a731cb52856dd9fe2/packages/client/README.md) page from the preview.

### CODE REVIEW

File: packages/client/examples/light-browser-sync.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
+# Example: Light Browser Sync
+
+## Light sync 
+
+In this example, we will run two ethereumjs-clients. The first will be a full sync client that
```
Comment (gabrocheleau) [inline packages/client/examples/light-browser-sync.md:5]:
```suggestion
In this example, we will run two ethereumjs clients. The first will be a full sync client that
```

File: packages/client/examples/light-browser-sync.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+will connect to the rinkeby network and start downloading the blockchain. The second will be a
```
Comment (gabrocheleau) [inline packages/client/examples/light-browser-sync.md:6]:
```suggestion
connects to the rinkeby network and starts downloading the blockchain. The second will be a
```

File: packages/client/examples/light-browser-sync.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+light client that connects to the first client and syncs headers as they are downloaded.
+
+The first client will use RLPx to connect to the rinkeby network, but will also provide a libp2p
+listener. The second client will use libp2p to connect to the first client.
+
+Run the first client and start downloading blocks:
+
+```
+ethereumjs --syncmode full --lightserv true  --datadir first --network rinkeby --transports rlpx libp2p --multiaddrs /ip4/127.0.0.1/tcp/50505/ws
```
Comment (gabrocheleau) [inline packages/client/examples/light-browser-sync.md:15]: Should we also include the bash instructions for when the client is run from the monorepo? i.e.:
```
npm run client:start  -- --syncmode full --lightserv true  --datadir first --network rinkeby --transports rlpx libp2p --multiaddrs /ip4/127.0.0.1/tcp/50505/ws
```
I think new users might be confused by the docs switching between the `ethereumjs` and `npm run client:start` commands

**Review by gabrocheleau — CHANGES_REQUESTED** [review]

File: packages/client/README.md [hunk 4]
```diff
@@ -1,29 +1,38 @@
...
+| --- |
 
-See [Technical Guidelines](#technical-guidelines) to dive directly into development info.
+Note: there is no current release on npm and the releases from the [standalone repository](https://github.com/ethereumjs/ethereumjs-client) are outdated. Use the latest `master` to run the client. There will be an up-to-date client release soon (Fall 2021) under a scoped `@ethereumjs/client` naming scheme. 
 
-Current development stage: `EARLY DEVELOPMENT`
+# INTRODUCTION
 
-# PROJECT SUMMARY
+The EthereumJS Client is an Ethereum Execution Client (similar to [go-ethereum](https://github.com/ethereum/go-ethereum) or [Nethermind](https://github.com/NethermindEth/nethermind)) written in `TypeScript`/`JavaScript`, the non-Smart-Contract language Ethereum dApp developers are most familiar with. It is targetet to be a client for research and development and not meant to be used in production on `mainnet` for the forseable future (out of ressource and security considerations).
```
Comment (ryanio) [inline packages/client/README.md:16]:
```suggestion
The EthereumJS Client is an Ethereum Execution Client (similar to [go-ethereum](https://github.com/ethereum/go-ethereum) or [Nethermind](https://github.com/NethermindEth/nethermind)) written in `TypeScript`/`JavaScript`, the non-Smart-Contract language Ethereum dApp developers are most familiar with. It is targeted to be a client for research and development and not meant to be used in production on `mainnet` for the foreseeable future (out of resource and security considerations). 
```

File: packages/client/README.md [hunk 5]
```diff
@@ -215,92 +235,51 @@ Output:
...
+#### Using Debug Loggers
+
+The client's logging verbosity level can be set with `--loglevel`.  Available levels are
+`error`, `warn`, `info`, `debug`.
+
+```shell
+ethereumjs --loglevel=debug
+```
+
+If you want to have verbose logging output across the stack you can use the f
```
Comment (ryanio) [inline packages/client/README.md:266]: unfinished sentence?

File: packages/client/examples/private-geth-network.md [hunk 6]
```diff
@@ -0,0 +1,40 @@
+# Example: Private Geth Network
+
+## Running a private network with geth
+
+In this example, we import the genesis parameters for a private Proof-of-Authority network using the geth genesis parameters format and then sync our client with a geth instance that is sealing blocks.
+
+First, create a signer account and place in the `data` directory to use with geth to seal blocks following [these instructions](https://geth.ethereum.org/docs/interface/managing-your-accounts).
+
+Next, open [these genesis parameters](./test/testdata/poa.json) and replace "728bb68502bfcd91ce4c7a692a0c0773ced5cff" with your signer address in both the `extradata` property and in the `alloc` section.
```
Comment (ryanio) [inline packages/client/examples/private-geth-network.md:9]: noticed this typo, missing a zero, @acolytec3 this is probably where you got stuck today

```suggestion
Next, open [these genesis parameters](./test/testdata/poa.json) and replace "728bb68502bfcd91ce4c7a692a0c0773ced5cff0" with your signer address in both the `extradata` property and in the `alloc` section.
```

File: packages/client/README.md [hunk 7]
```diff
@@ -32,85 +41,96 @@ For the `ethereumjs` CLI command to work run:
...
-ethereumjs --network=mainnet [--loglevel=debug]
-```
-### CLI reference [WORK-IN-PROGRESS] 
+## USAGE
 
-#### `--network`
-You can connect to specific networks by name using the `--network` parameter as below: (`rinkeby` is used here)
+### Introduction
+
+You can get up the client up and running by going to the shell and run:
```
Comment (ryanio) [inline packages/client/README.md:56]:
```suggestion
You can get the client up and running by going to the shell and run:
```

File: packages/client/examples/light-browser-sync.md (hunk 3 above)
Comment (ryanio) [reply packages/client/examples/light-browser-sync.md:15]: yeah I agree, the new beginning of the README does explain this a bit, though for the examples i think we should stick to one or the other and not switch between.

File: packages/client/README.md (hunk 5 above)
Comment (ryanio) [inline packages/client/README.md:266]:
```suggestion
```

File: packages/client/README.md [hunk 8]
```diff
@@ -215,92 +235,51 @@ Output:
...
+The client's logging verbosity level can be set with `--loglevel`.  Available levels are
+`error`, `warn`, `info`, `debug`.
+
+```shell
+ethereumjs --loglevel=debug
+```
+
+If you want to have verbose logging output across the stack you can use the f
+
+For more in-depth debugging on networking the the underlying [devp2p](../devp2p) library integrates with the [debug](https://github.com/visionmedia/debug) package and can also be used from within a client execution context:
```
Comment (ryanio) [inline packages/client/README.md:268]:
```suggestion
For more in-depth debugging on networking the underlying [devp2p](../devp2p) library integrates with the [debug](https://github.com/visionmedia/debug) package and can also be used from within a client execution context:
```

File: packages/client/README.md [hunk 9]
```diff
@@ -215,92 +235,51 @@ Output:
...
+For an overview on the design goals which served as a guideline on design decisions as well as some structural client overview see the dedicated [DESIGN.md](./DESIGN.md) document.
 
-**EthereumJS Ecosystem**
+### Client Customization
 
-This project will be embedded in the EthereumJS ecosystem and many submodules already exist and
-can be used within the project, have a look e.g. at [@ethereumjs/block](https://github.com/ethereumjs/ethereumjs-monorepo/tree/master/packages/block), [@ethereumjs/vm](https://github.com/ethereumjs/ethereumjs-monorepo/tree/master/packages/vm),
-[merkle-patricia-tree](https://github.com/ethereumjs/ethereumjs-monorepo/tree/master/packages/trie) or the
-[@ethereumjs/devp2p](https://github.com/ethereumjs/ethereumjs-monorepo/tree/master/packages/devp2p) implementation. Work needs to be done both within these repos and related libraries.
+For getting a start on customizing the client and use programatically see the code from [./bin/cli.ts](./bin/cli.ts) to get an idea how an [EthereumClient](./lib/client.ts) instance is invoked programatically.
```
Comment (ryanio) [inline packages/client/README.md:247]:
```suggestion
To get a start on customizing the client and using it programmatically see the code from [./bin/cli.ts](./bin/cli.ts) to get an idea of how an [EthereumClient](./lib/client.ts) instance is invoked programmatically.
```

File: packages/client/examples/light-browser-sync.md (hunk 3 above)
Comment (ryanio) [reply packages/client/examples/light-browser-sync.md:15]:
```suggestion
npm run client:start -- --syncmode full --lightserv true  --datadir first --network rinkeby --transports rlpx libp2p --multiaddrs /ip4/127.0.0.1/tcp/50505/ws
```

File: packages/client/examples/light-browser-sync.md [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+INFO [10-24|11:37:48] Listener up transport=libp2p url=<b>/ip4/127.0.0.1/tcp/50505/ws/p2p/QmYAuYxw6QX1x5aafs6g3bUrPbMDifP5pDun3N9zbVLpEa</b>
+...
+</pre>
+
+Copy the libp2p URL from the output. In this example, the url is `/ip4/127.0.0.1/tcp/50505/ws/p2p/QmYAuYxw6QX1x5aafs6g3bUrPbMDifP5pDun3N9zbVLpEa` but it will be different in your case.
+
+Wait until a few thousand blocks are downloaded and then run the second client in a new terminal, using the url above to connect to the first client:
+
+<pre>
+ethereumjs --syncmode light --network rinkeby --datadir second --transports libp2p --multiaddrs /ip4/0.0.0.0/tcp/50506 --bootnodes=<b>/ip4/127.0.0.1/tcp/50505/ws/p2p/QmYAuYxw6QX1x5aafs6g3bUrPbMDifP5pDun3N9zbVLpEa</b>
```
Comment (ryanio) [inline packages/client/examples/light-browser-sync.md:32]:
```suggestion
npm run client:start -- --syncmode light --network rinkeby --datadir second --transports libp2p --multiaddrs /ip4/0.0.0.0/tcp/50506 --bootnodes=<b>/ip4/127.0.0.1/tcp/50505/ws/p2p/QmYAuYxw6QX1x5aafs6g3bUrPbMDifP5pDun3N9zbVLpEa</b>
```

File: packages/client/README.md [hunk 11]
```diff
@@ -32,85 +41,96 @@ For the `ethereumjs` CLI command to work run:
...
-DEBUG=*,-babel [CLIENT_START_COMMAND]
+ethereumjs --network=rinkeby
 ```
 
-for all output or something more targeted by listing the loggers like
+The client currently supports `full` sync being set as a default and has experimental support for `light` sync.
+
+### Custom Chains
+
+The EthereumJS client supports running custom chains based on a custom chain configuration. There are two ways of reading in custom chain configuration parameters
```
Comment (ryanio) [inline packages/client/README.md:103]:
```suggestion
The EthereumJS client supports running custom chains based on a custom chain configuration. There are two ways of reading in custom chain configuration parameters:
```

File: packages/client/DESIGN.md [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+  other components as possible. This reduces dependencies between PRs and encourages contributors
+  to work in parallel. It also improves extensibility of the code as new features like sharding
+  and libp2p support are added.
+- **Easily tested**: The design should make testing of individual components as easy as possible.
+  This goes hand in hand with the previous goal of loose coupling.
+- **Readable code**: More readable code should encourage more contributions from the community and help
+  with bug fixing.
+- **Well documented**: Similar to above, this will help both contributors and users of the project.
+
+The current design tries to achieves the goals of loose coupling and ease of testing by using an
```
Comment (ryanio) [inline packages/client/DESIGN.md:17]:
```suggestion
The current design tries to achieve the goals of loose coupling and ease of testing by using an
```

File: packages/client/README.md [hunk 13]
```diff
@@ -1,29 +1,38 @@
...
+# @ethereumjs/client
 
 [![NPM Package][client-npm-badge]][client-npm-link]
 [![GitHub Issues][client-issues-badge]][client-issues-link]
 [![Actions Status][client-actions-badge]][client-actions-link]
 [![Code Coverage][client-coverage-badge]][client-coverage-link]
 [![Discord][discord-badge]][discord-link]
 
-This is the work repository for the EthereumJS client project targeting both Node.js and the browser.
+| Ethereum Execution (Eth 1.0) Client build in TypeScript/JavaScript. |
```
Comment (ryanio) [inline packages/client/README.md:9]:
```suggestion
| Ethereum Execution (Eth 1.0) Client built in TypeScript/JavaScript. |
```

File: packages/client/DESIGN.md [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+
+**Directory structure**
+
+- `/bin` Contains the CLI script for the `ethereumjs` command.
+- `/docs` Contains auto-generated API docs.
+- `/lib/blockchain` Contains the `Chain` class.
+- `/lib/net` Contains all of the network layer classes including `Peer`, `Protocol` and its subclasses, `Server` and its subclasses, and `PeerPool`.
+- `/lib/service` Contains the main Ethereum services (`FullEthereumService` and `LightEthereumService`).
+- `/lib/rpc` Contains the RPC server (optionally) embedded in the client.
+- `/lib/sync` Contains the various chain synchronizers and `Fetcher` helpers.
```
Comment (ryanio) [inline packages/client/DESIGN.md:35]:
```suggestion
- `/lib/sync` Contains the various chain synchronizers and `Fetcher` helpers.
- `/lib/miner` Contains the miner module that can build new blocks on top of the chain.
```

File: packages/client/DESIGN.md [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+
+- `Chain` [**In Progress**] This class represents the blockchain and is a wrapper around
+  `@ethereumjs/blockchain`. It handles creation of the data directory, provides basic blockchain operations
+  and maintains an updated current state of the blockchain, including current height, total difficulty, and
+  latest block.
+- `Server` This class represents a server that discovers new peers and handles incoming and dropped
+  connections. When a new peer connects, the `Server` class will negotiate protocols and emit a `connected`
+  event with a new `Peer`instance. The peer will have properties corresponding to each protocol. For example,
+  if a new peer understands the `eth` protocol, it will contain an `eth` property that provides all `eth`
+  protocol methods (for example: `peer.eth.getBlockHeaders()`) - `RlpxServer` [**In Progress**] Subclass of `Server` that implements the `devp2p/rlpx` transport. - `Libp2pServer` [**In Progress**] Subclass of `Server` that implements the `libp2p` transport.
```
Comment (ryanio) [inline packages/client/DESIGN.md:48]:
```suggestion
  protocol methods (for example: `peer.eth.getBlockHeaders()`)
    - `RlpxServer` [**In Progress**] Subclass of `Server` that implements the `devp2p/rlpx` transport.
    - `Libp2pServer` [**In Progress**] Subclass of `Server` that implements the `libp2p` transport.
```

File: packages/client/DESIGN.md [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+- `Peer` Represents a network peer. Instances of `Peer` are generated by the `Server`
+  subclasses and contain instances of supported protocol classes as properties. Instances of `Peer` subclasses can also be used to directly connect to other nodes via the `connect()` method. Peers emit `message` events
+  whenever a new message is received using any of the supported protocols. - `RlpxPeer` [**In Progress**] Subclass of `Peer` that implements the `devp2p/rlpx` transport. - `Libp2pPeer` [**In Progress**] Subclass of `Peer` that implements the `libp2p` transport.
```
Comment (ryanio) [inline packages/client/DESIGN.md:51]:
```suggestion
  whenever a new message is received using any of the supported protocols.
    - `RlpxPeer` [**In Progress**] Subclass of `Peer` that implements the `devp2p/rlpx` transport.
    - `Libp2pPeer` [**In Progress**] Subclass of `Peer` that implements the `libp2p` transport.
```

File: packages/client/DESIGN.md [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+- `PeerPool` [**In Progress**] Represents a pool of network peers. `PeerPool` instances emit `added`
+  and `removed` events when new peers are added and removed and also emit the `message` event whenever
+  any of the peers in the pool emit a message. Each `Service` has an associated `PeerPool` and they are used primarily by `Synchronizer`s to help with blockchain synchronization.
+- `Synchronizer` Subclasses of this class implements a specific blockchain synchronization strategy. They
+  also make use of subclasses of the `Fetcher` class that help fetch headers and bodies from pool peers. The fetchers internally make use of streams to handle things like queuing and backpressure. - `FullSynchronizer` [**In Progress**] Implements full syncing of the blockchain - `LightSynchronizer` [**In Progress**] Implements light syncing of the blockchain
```
Comment (ryanio) [inline packages/client/DESIGN.md:58]:
```suggestion
  also make use of subclasses of the `Fetcher` class that help fetch headers and bodies from pool peers. The fetchers internally make use of streams to handle things like queuing and backpressure.
    - `FullSynchronizer` [**In Progress**] Implements full syncing of the blockchain
    - `LightSynchronizer` [**In Progress**] Implements light syncing of the blockchain
```

File: packages/client/DESIGN.md [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+  also make use of subclasses of the `Fetcher` class that help fetch headers and bodies from pool peers. The fetchers internally make use of streams to handle things like queuing and backpressure.
+    - `FullSynchronizer` [**In Progress**] Implements full syncing of the blockchain
+    - `LightSynchronizer` [**In Progress**] Implements light syncing of the blockchain
+- `Handler` Subclasses of this class implements a protocol message handler. Handlers respond to incoming requests from peers.
+  - `EthHandler` [**In Progress**] Handles incoming ETH requests
+  - `LesHandler` [**In Progress**] Handles incoming LES requests
+- `Service` Subclasses of `Service` will implement specific functionality of a `Client`. For example, the `EthereumService` subclasses will synchronize the blockchain using the full or light sync protocols. Each service must specify which protocols it needs and define a `start()` and `stop()` function.
+  - `FullEthereumService` [**In Progress**] Implementation of ethereum full sync.
+  - `LightEthereumService` [**In Progress**] Implementation of ethereum light sync.
+  - `WhisperService` [**Not Started**] Implementation of an ethereum whisper node.
```
Comment (ryanio) [inline packages/client/DESIGN.md:74]:
```suggestion
```

**Review by ryanio — APPROVED** [review]
lgtm, awesome improvements!

**Comment by holgerd77** [comment]
@ryanio thanks a lot for these extensive continued iterations and bringing this over the finish line!! 🌻 😀

**Pull request merged**